### PR TITLE
SMB Features

### PR DESCRIPTION
**Pull request by mbscit**

### CODE REVIEW

File: smb/src/main/java/ch/cyberduck/core/smb/SMBReadFeature.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                }
+                catch(IOException e) {
+                    fileEntry.close();
+                    throw new BackgroundException(e);
+                }
+            }
+
+            return new SMBInputStream(stream, fileEntry);
+        } catch(SMBApiException e) {
+            throw new SmbExceptionMappingService().map(e);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBReadFeature.java:70]:
```suggesetion
            throw new SmbExceptionMappingService().map("Download {0} failed", e, file);
```

File: smb/src/main/java/ch/cyberduck/core/smb/SMBWriteFeature.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            createOptions.add(SMB2CreateOptions.FILE_NON_DIRECTORY_FILE);
+
+
+            File fileEntry = session.share.openFile(file.getAbsolute(), accessMask, fileAttributes,
+                    shareAccessSet, smb2CreateDisposition, createOptions);
+
+            return new VoidStatusOutputStream(new SMBOutputStream(fileEntry.getOutputStream(), fileEntry));
+        }
+        catch(SMBApiException e) {
+            throw new SmbExceptionMappingService().map(e);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBWriteFeature.java:58]:
```suggestion
            throw new SmbExceptionMappingService().map("Upload {0} failed", e, file);
```

File: smb/src/main/java/ch/cyberduck/core/smb/SMBAttributesFinderFeature.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            attributes.setAccessedDate(fileInformation.getBasicInformation().getLastAccessTime().toEpochMillis());
+            attributes.setModificationDate(fileInformation.getBasicInformation().getLastWriteTime().toEpochMillis());
+            attributes.setCreationDate(fileInformation.getBasicInformation().getCreationTime().toEpochMillis());
+            attributes.setSize(fileInformation.getStandardInformation().getEndOfFile());
+            attributes.setDisplayname(fileInformation.getNameInformation().substring(1));
+
+            return attributes;
+        }
+        catch(SMBApiException e) {
+            throw new SmbExceptionMappingService().map(e);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBAttributesFinderFeature.java:44]:
```suggestion
            throw new SmbExceptionMappingService().map("Failure to read attributes of {0}", e, file);
```

File: smb/src/main/java/ch/cyberduck/core/smb/SMBReadFeature.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        }
+    }
+
+    public final class SMBInputStream extends InputStream {
+
+        private InputStream stream;
+        private File file;
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBReadFeature.java:77]: Add `final`

File: smb/src/main/java/ch/cyberduck/core/smb/SMBWriteFeature.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+
+    }
+
+    public final class SMBOutputStream extends OutputStream {
+
+        private OutputStream stream;
+        private File file;
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBWriteFeature.java:67]: Add `final`

**Review by dkocher — CHANGES_REQUESTED** [review]

File: smb/src/main/java/ch/cyberduck/core/smb/SMBCopyFeature.java [hunk 6]
```diff
@@ -96,16 +82,17 @@ private Path copyFile(Path source, Path target, TransferStatus status, Connectio
             sourceFile.remoteCopyTo(targetFile);
         }
         catch(TransportException | BufferException e) {
+            throw new BackgroundException(e);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBCopyFeature.java:85]: Replace with parameters `"Cannot copy {0}", e, source`

```suggestion
            throw new BackgroundException("Cannot copy {0}", e, source);
```

File: smb/src/main/java/ch/cyberduck/core/smb/SMBCopyFeature.java [hunk 7]
```diff
@@ -44,31 +42,19 @@ public Path copy(Path source, Path target, TransferStatus status, ConnectionCall
             return target;
         }
         catch(SMBApiException e) {
-            throw new SmbExceptionMappingService().map(e);
+            throw new SMBExceptionMappingService().map(e);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBCopyFeature.java:45]: Replace with parameters `"Cannot copy {0}", e, source`

```suggestion
            throw new BackgroundException("Cannot copy {0}", e, source);
```

File: smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java [hunk 8]
```diff
@@ -70,13 +70,13 @@ public AttributedList<Path> list(final Path directory, final ListProgressListene
...
 
-                result.add(new Path(directory, fileName, type).withAttributes(attributes));
+                result.add(new Path(directory, fileName, type, attributes));
 
             }
             return result;
         }
         catch(SMBApiException e) {
-            throw new SmbExceptionMappingService().map(e);
+            throw new SMBExceptionMappingService().map(e);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBListService.java:79]: Replace with parameters `"Listing directory {0} failed", e, directory`

```suggestion
            throw new SMBExceptionMappingService().map("Listing directory {0} failed", e, directory);
```

**Review by dkocher — CHANGES_REQUESTED** [review]

File: smb/src/main/java/ch/cyberduck/core/smb/SMBTimestampFeature.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+        try (File fileEntry = session.share.openFile(file.getAbsolute(), accessMask, fileAttributes, shareAccessSet, smb2CreateDisposition, createOptions)) {
+            FileTime creationTime = fileEntry.getFileInformation().getBasicInformation().getCreationTime();
+            FileTime time = FileTime.ofEpochMillis(status.getTimestamp());
+
+            FileBasicInformation fileBasicInformation = new FileBasicInformation(creationTime, time, time, time, FileAttributes.FILE_ATTRIBUTE_NORMAL.getValue());
+            fileEntry.setFileInformation(fileBasicInformation);
+        }
+        catch(SMBApiException e) {
+            throw new SMBExceptionMappingService().map("Setting timestamp on file {0} failed", e, file);
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBTimestampFeature.java:54]: Use localized text `Cannot change timestamp of {0}`

```suggestion
            throw new SMBExceptionMappingService().map("Cannot change timestamp of {0}", e, file);
```

File: smb/src/main/java/ch/cyberduck/core/smb/SMBProtocol.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+import ch.cyberduck.core.AbstractProtocol;
+import ch.cyberduck.core.Credentials;
+import ch.cyberduck.core.LocaleFactory;
+import ch.cyberduck.core.LoginOptions;
+import ch.cyberduck.core.Scheme;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class SMBProtocol extends AbstractProtocol {
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBProtocol.java:26]: Override `getUsernamePlaceholder` to provide placeholder text with username format requiring share name.

File: smb/src/main/java/ch/cyberduck/core/smb/SMBReadFeature.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+                }
+            }
+
+            return new SMBInputStream(stream, fileEntry);
+        } catch(SMBApiException e) {
+            throw new SMBExceptionMappingService().map("Download {0} failed", e, file);
+        }
+    }
+
+    private final class SMBInputStream extends ProxyInputStream {
```
Comment (dkocher) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBReadFeature.java:75]: Make static.

**Review by dkocher — CHANGES_REQUESTED** [review]
- [ ] Sqash commits
- [ ] Review comments
- [ ] Resolve TODOs

**Review by dkocher — CHANGES_REQUESTED** [review]
- [ ] Sqash commits
- [ ] Review comments
- [ ] Resolve TODOs

**Comment by mbscit** [comment]
We have opened a new PR with all comments resolved and the commits squashed
[https://github.com/iterate-ch/cyberduck/pull/15006](https://github.com/iterate-ch/cyberduck/pull/15006
)

**Pull request closed without merging**